### ISSUE TITLE
App crashes when accessing note and go back to list

### ISSUE THREAD

**Issue by stefan-niedermann** [stefan-niedermann/nextcloud-notes#199]
It is very hard to reproduce this issue. Appears with current master and v0.10.0.

Steps:

1. Open any note from the list view
2. Close this note as quick as possible (using native back-button or top left arrow button does not matter)
3. Repeat Steps 1 - 2 until the app crashes (10 - 15 times at my device)

CC @korelstar 

```
03-31 08:55:11.909 2691-2725/? D/SyncTask:    Process Remote Note: R20  (*)  / MarkDown Test (2017-03-22 20:08:07 / null)
03-31 08:55:11.910 2691-2725/? D/SyncTask:    ... found -> Update
                                           
                                           --------- beginning of crash
03-31 08:55:11.915 2691-2725/? E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #3
                                                 Process: it.niedermann.owncloud.notes, PID: 2691
                                                 java.lang.RuntimeException: An error occurred while executing doInBackground()
                                                     at android.os.AsyncTask$3.done(AsyncTask.java:325)
                                                     at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
                                                     at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
                                                     at java.util.concurrent.FutureTask.run(FutureTask.java:242)
                                                     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
                                                     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                                                     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                     at java.lang.Thread.run(Thread.java:761)
                                                  Caused by: java.lang.IllegalStateException: attempt to re-open an already-closed object: SQLiteDatabase: /data/user/0/it.niedermann.owncloud.notes/databases/OWNCLOUD_NOTES
                                                     at android.database.sqlite.SQLiteClosable.acquireReference(SQLiteClosable.java:55)
                                                     at android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java:1548)
                                                     at android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java:1525)
                                                     at it.niedermann.owncloud.notes.persistence.NoteSQLiteOpenHelper.updateNote(NoteSQLiteOpenHelper.java:407)
                                                     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.pullRemoteChanges(NoteServerSyncHelper.java:278)
                                                     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:201)
                                                     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:170)
                                                     at android.os.AsyncTask$2.call(AsyncTask.java:305)
                                                     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243) 
                                                     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                     at java.lang.Thread.run(Thread.java:761) 
```

**Comment by korelstar** [comment]
Hmmm. That's not cool. Who closes the database, here? In the code, the `close()` is after the respective line: https://github.com/stefan-niedermann/nextcloud-notes/blob/4afb170adad86936f915bc658e2cf5f2678ed34f/app/src/main/java/it/niedermann/owncloud/notes/persistence/NoteSQLiteOpenHelper.java#L407

At first glance, I can't see anything wrong, there. Is another thread able to close the database connection? But this is only possible, if `getWritableDatabase()` provides the same database connection for multiple threads. In my view, this *should* not be true.

Sorry, I have no ideas, currently.

**Comment by fabienli** [comment]
I have the same issue happening here, and with ~250 notes, this happens quite often.
From logs, I understand that when leaving EditNoteActivity instance, saveData(...) is called while a background sync is ongoing. adding some logs here and there to dump db.hashCode(), getWritableDatabase() seems to actually send the same db reference twice at some point.
I've tried updating append getWritableDatabase() calls in NoteSQliteOpenHelper.java with:
        SQLiteDatabase db = this.getWritableDatabase();
        db.acquireReference();
and the issue doesn't seem to happen anymore, as the db.close() now waits for all references to be released before actually closing.
Let me know what you think of this solution?

**Comment by korelstar** [comment]
Thanks for your contribution (also in #203)!

Honestly, the method `acquireReference()` is new to me. I had to look into the Android sources (`android.database.sqlite.SQLiteClosable`) in order to understand it. This method has the purpose to count references to the `db` object. The database is closed only if there are no more references on that object. For this purpose, it has the private variable `mReferenceCount` that is set to `1` on instantiation, increased by `1` when calling `acquireReference()` and decreased by `1` when calling `releaseReference()`. `close()` is the same as `releaseReference()`: the `SQLiteClosable` is closed only if `mReferenceCount=0` (after decreasing the counter).

However, I'm wondering if this whole thing is really needed, since I can't find any use of these methods in any other apps. But I found an interesting article, where concurrent use of a database is discussed: https://nfrolov.wordpress.com/2014/08/16/android-sqlitedatabase-locking-and-multi-threading/
There, it is recommended to just omit closing the database. Until now, I thought that Android will give me warning if I do so, but in my tests, everything works fine! (Weird) Does this approach work for you, too?
